### ISSUE TITLE
order of run phase when slicing is on?

### ISSUE THREAD

**Issue by amirhosseinvz94** [openresty/lua-nginx-module#1483]
in openresty 1.15.x version , when slicing is on (like: slice 100m ;) for origin content (big than 100 m: 1gb) with wget, in first request range[0-100] openresty phases run order is correct and start from Rewrite/Access phase, but in seccond request range [100-200 ] openresty phase start from Content_phase (balancer_by_lua), but i need ctx that set in older phase and perevius reqest range [0-100] context is cleand , and in this phase and i cant get it because not access to connect redis.
in old version like 1.11.x or 1.13.x every thing is correct , but in 1.15.x version start from Content_phase and is a problem
What will i do؟

**Comment by agentzh** [comment]
We never touched the slice module in OpenResty nor did we touch the nginx subrequest model. So it looks like a recent changes in new versions of the standard slice module.

From your description, it seems like slice module explicitly skip the rewrite phase in its subrequests (note that subrequests never run the access phase at all). To be sure, you can temporarily enable the nginx debug logs in your openresty build (or just use the openresty-debug rpm/deb pre-built packages if you are using our binary packages) and see what's going on in all those `[debug]` level error messages produced by the nginx internals.

**Comment by amirhosseinvz94** [comment]
thank you so much, 
this line in older version of slice module :
    if (ngx_http_subrequest(r, &r->uri, &r->args, &sr, NULL, 0) != NGX_OK) {
        return NGX_ERROR;
    }
changed to :
    if (ngx_http_subrequest(r, &r->uri, &r->args, &ctx->sr, NULL,
                            NGX_HTTP_SUBREQUEST_CLONE)
in new version, NGX_HTTP_SUBREQUEST_CLONE flag Causes to subrequest inheritt phase_handler and  content_handler
this part of subrequest logic is :  
if (flags & NGX_HTTP_SUBREQUEST_CLONE) {
        sr->method = r->method;
        sr->method_name = r->method_name;
        sr->loc_conf = r->loc_conf;
        sr->valid_location = r->valid_location;
        sr->valid_unparsed_uri = r->valid_unparsed_uri;
        sr->content_handler = r->content_handler;
        sr->phase_handler = r->phase_handler;
        sr->write_event_handler = ngx_http_core_run_phases;

**Comment by agentzh** [comment]
@amirhosseinvz94 Thanks for digging it up then it seems the cloned subrequests indeed skip earlier phases. It looks like an intentional optimization.

**Comment by agentzh** [comment]
I don't think we can do much here, right? It's by design.

**Comment by amirhosseinvz94** [comment]
Yes this is part of the nginx core and making changes in it wouldn't be a logical idea. But I think the best solution would be if we could have some customizable settings in openresty so that we could share the data(ctx) between the request and subrequest

**Comment by agentzh** [comment]
@amirhosseinvz94 The problem is that those subrequests are not created nor managed by OpenResty's nginx modules and the nginx core does not provide hooks to monitor all subrequests created by other nginx C modules. So to do what you want, we'll have to patch the nginx core anyway.

**Comment by amirhosseinvz94** [comment]
thanks ,this patch could be so much useful and usecase, like internal redirect happened in error_page, that we have to connect redis again for getting data and this scenario has overhead

**Comment by agentzh** [comment]
@amirhosseinvz94 No, you could cache the metadata in shdict so that you don't need to talk to redis again (as long as you can predict what data you will need).
